src/puppetlabs/puppetdb/cli/benchmark.clj

```diff
@@ -331,34 +313,43 @@
                                   [data-paths false])]
       (kitchensink/mapvals #(some-> % (load-sample-data from-cp?)) data-paths))))
 
-(defn try-read-and-send-command [base-url command-send-ch]
-  (try
-    (cm/match [(<!! command-send-ch)]
-              [:stop] nil
-              [nil] nil
-              [[entity host version payload]]
-              (do ((case entity
-                     :catalog client/submit-catalog
-                     :report client/submit-report
-                     :factset client/submit-facts)
-                   base-url host version payload)
-                  ::submitted))
-    (catch Exception e
-      (println "Exception while submitting command: " e)
-      ::error)))
-
 (defn start-command-sender
-  "Start a command sending thread. Reads commands from command-send-ch of the
-  form [entity version payload-string]. Writes a value to rate-monitor-ch for
-  every command sent."
-  [base-url command-send-ch rate-monitor-ch]
-  (future
-    (loop []
-      (case (try-read-and-send-command base-url command-send-ch)
-        ::submitted (do (>!! rate-monitor-ch true)
-                        (recur))
-        ::error (recur)
-        nil nil))))
+  "Start a command sending process in the background. Reads host-state maps from
+  command-send-ch and sends commands to the puppetdb at base-url. Writes
+  ::submitted to rate-monitor-ch for every command sent, or ::error if there was
+  a problem. Close command-send-ch to stop the background process."
+  [base-url command-send-ch rate-monitor-ch num-threads]
+  (let [fanout-commands-ch (chan)]
+    ;; fanout: given a single host state, emit 3 messages, one for each command.
+    ;; This gives better parallelism for message submission.
+    (async/pipeline
+     1
+     fanout-commands-ch
+     (mapcat (fn [host-state]
+               (let [{:keys [host catalog report factset]} host-state]
+                 (remove nil?
+                         [(when catalog [:catalog host 9 catalog])
+                          (when report [:report host 8 report])
+                          (when factset [:factset host 5 factset])]))))
+     command-send-ch)
+
+    ;; actual sender process
+    (async/pipeline-blocking
+     num-threads
+     rate-monitor-ch
+     (map (fn [[command host version payload]]
+            (let [submit-fn (case command
+                              :catalog client/submit-catalog
+                              :report client/submit-report
+                              :factset client/submit-facts)]
+
+              (try
+                (submit-fn base-url host version payload)
+                ::submitted
```

Comment:
It's not. I'm not sure it ever was, actually. I'm kind of loathe to remove it, as the rate monitor needs /something/ to count.